src/gtest/test_blockdownload.cpp

```diff
@@ -0,0 +1,74 @@
+#include <gtest/gtest.h>
+#include <gmock/gmock.h>
+
+#include "main.h"
+
+using namespace zen;
+
+// Checks IsInitialBlockDownload() conditions and lockIBDState latching to false
+TEST(initialblockdwnld, checkIBDState) {
+    SelectParams(CBaseChainParams::MAIN);
+    const CChainParams& chainParams = Params();
+
+    // 1.
+    // fImporting, fReindex and fReindexFast initialized as false
+    // Check each individual variable and restore its state
+    fImporting   = true; EXPECT_TRUE(IsInitialBlockDownload()); fImporting   = false;
+    fReindex     = true; EXPECT_TRUE(IsInitialBlockDownload()); fReindex     = false;
+    fReindexFast = true; EXPECT_TRUE(IsInitialBlockDownload()); fReindexFast = false;
+
+    // 2.
+    // Expecting that fImporting, fReindex and fReindexFast are false, check if
+    // chainActive.Height() < Checkpoints::GetTotalBlocksEstimate(chainParams.Checkpoints())
+    fCheckpointsEnabled = true;
+    EXPECT_TRUE(IsInitialBlockDownload());
+
+    // Setting tip of active chain lower than checkpoint
+    auto block1 = CBlockIndex();
+    block1.nHeight = Checkpoints::GetTotalBlocksEstimate(chainParams.Checkpoints()) - 1;
+    chainActive.SetTip(&block1);
+    EXPECT_TRUE(IsInitialBlockDownload());
```

Comment:
Would it be possible to include this section in a loop and check that we always get `true` for every height in [0, 1, 2, ..., N] up to the checkpoint height?

I'm asking because in this way I guess we can spot also weird behaviors, like the one we had with the call to `isAfterChainsplit()` inside `IsInitialBlockDownload()`.

Reply:
The test has been updated for checking every value up to height N.